### PR TITLE
[Profiler] Unify global and thread local profiler lookup.

### PR DESCRIPTION
**Pull request by robieta**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84276
* #84275
* #84274
* #84273
* #83965
* #83964
* #83895
* __->__ #83894
* #83893
* #83892

This PR renames `ProfilerThreadLocalStateBase` to simply `ProfilerStateBase`, and adds `push`, `pop`, and `get` methods. `global` can be specified, or can be omitted for priority selection.

In order to support this unification it was necessary to make a (mostly) non-throwing version of pop. The asserts around observer removal are intended to act as guard rails against multiple profilers trampling over each other. However on-demand wants to do exactly that because it wants to be able to preempt.

A hack would be to get the current observer and then only pop if an observer is found, but that would be prone to race conditions. By removing the asserts, we can preserve the old behavior by adding `ASSERT(pop())` on the caller side while allowing more complex handling for the kineto client interface. (Later PR.)

Differential Revision: [D38931521](https://our.internmc.facebook.com/intern/diff/D38931521/)

### CODE REVIEW

**Review by slgong-fb — APPROVED** [review]
LGTM overall.

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 1]
```diff
@@ -200,20 +201,21 @@ static inline uint64_t getForwardThreadKey(uint64_t tid, uint64_t seqNr) {
...
-      : ProfilerThreadLocalStateBase(config),
+      : ProfilerStateBase(config),
         start_time_(getTimeUs()),
         record_queue_(config, activities) {}
   ~KinetoThreadLocalState() override = default;
 
-  static KinetoThreadLocalState* getTLS() {
-    auto tls = ProfilerThreadLocalStateBase::getTLS();
+  static KinetoThreadLocalState* get(bool global) {
+    auto* state = ProfilerStateBase::get(/*global=*/global);
```
[the remark this answers is not in the document]
Comment (robieta) [reply torch/csrc/autograd/profiler_kineto.cpp:214]: It's not a typo; it's to signal to the reader that `state` is a pointer type. (I'm trying to run down all the cases in profiler where this happens and replace `auto` with `auto*`) It also provides some extra compile time safety, as the compiler will yell if `get` returns a non-pointer type. (Including something like `shared_ptr`)

**Pull request closed without merging**